flatfs.go

```diff
@@ -1210,46 +1083,400 @@ func (fs *Datastore) Close() error {
 	return nil
 }
 
+// DiscardableBatch is an optional interface for batches that support discarding changes
+type DiscardableBatch interface {
+	datastore.Batch
+	Discard(ctx context.Context) error
+}
+
+// BatchReader is an optional interface for batches that support read operations
+type BatchReader interface {
+	datastore.Batch
+	datastore.Read
+}
+
 type flatfsBatch struct {
-	puts    map[datastore.Key][]byte
+	mu      sync.Mutex
+	puts    []datastore.Key
 	deletes map[datastore.Key]struct{}
 
-	ds *Datastore
+	ds        *Datastore
+	tempDir   string
+	tempFiles []string // track temp file names (without path) for cleanup
 }
 
 func (fs *Datastore) Batch(_ context.Context) (datastore.Batch, error) {
+	// Create a unique temp directory for this batch
+	tempDir := filepath.Join(fs.tempPath, fmt.Sprintf("batch-%d-%d", time.Now().UnixNano(), rand.Int63()))
+	if err := os.Mkdir(tempDir, 0755); err != nil {
+		return nil, fmt.Errorf("failed to create batch temp directory: %w", err)
+	}
+
 	return &flatfsBatch{
-		puts:    make(map[datastore.Key][]byte),
-		deletes: make(map[datastore.Key]struct{}),
-		ds:      fs,
+		puts:      make([]datastore.Key, 0),
+		deletes:   make(map[datastore.Key]struct{}),
+		ds:        fs,
+		tempDir:   tempDir,
+		tempFiles: make([]string, 0),
 	}, nil
 }
 
 func (bt *flatfsBatch) Put(ctx context.Context, key datastore.Key, val []byte) error {
 	if !keyIsValid(key) {
 		return fmt.Errorf("when putting '%q': %v", key, ErrInvalidKey)
 	}
-	bt.puts[key] = val
+
+	bt.mu.Lock()
+	defer bt.mu.Unlock()
+
+	// Check if batch temp directory exists, recreate if needed (for reuse after commit/discard)
+	if _, err := os.Stat(bt.tempDir); os.IsNotExist(err) {
```

Comment:
New code should use `errors.Is(err, fs.ErrNotExist)` instead of `os.IsNotExist(err)`. See: https://pkg.go.dev/os@go1.25.1#IsNotExist
```suggestion
	_, err := os.Stat(bt.tempDir)
	if errors.Is(err, fs.ErrNotExist) {
```